core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -1512,35 +1542,30 @@ class ParallelReduce< FunctorType
-                                                                 arg_policy.team_scratch_size(0),arg_policy.thread_scratch_size(0) ) /
-                                                                 arg_policy.vector_length() )
-    ), arg_policy.scratch_size(1,( 0 <= arg_policy.team_size() ? arg_policy.team_size() :
-        Kokkos::Impl::cuda_get_opt_block_size< ParallelReduce, LaunchBounds >( arg_functor , arg_policy.vector_length(),
-                                                                 arg_policy.team_scratch_size(0),arg_policy.thread_scratch_size(0) ) /
-                                                                 arg_policy.vector_length() )
-        )}
   , m_league_size( arg_policy.league_size() )
-  , m_team_size( 0 <= arg_policy.team_size() ? arg_policy.team_size() :
-      Kokkos::Impl::cuda_get_opt_block_size< ParallelReduce, LaunchBounds >( arg_functor , arg_policy.vector_length(),
-                                                               arg_policy.team_scratch_size(0),arg_policy.thread_scratch_size(0) ) /
-                                                               arg_policy.vector_length() )
+  , m_team_size( arg_policy.team_size() )
   , m_vector_size( arg_policy.vector_length() )
   {
+    cudaFuncAttributes attr = CudaParallelLaunch< ParallelReduce, LaunchBounds >::
+        get_cuda_func_attributes();
+    m_team_size = m_team_size>=0?m_team_size:
+        Kokkos::Impl::cuda_get_opt_block_size< FunctorType, LaunchBounds>(
+      m_policy.space().impl_internal_space_instance(),
+      attr, m_functor , m_vector_size,
+      m_policy.team_scratch_size(0), m_policy.thread_scratch_size(0) )/m_vector_size;
+
     // Return Init value if the number of worksets is zero
-    if( arg_policy.league_size() == 0) {
+    if( m_league_size*m_team_size == 0) {
       ValueInit::init( ReducerConditional::select(m_functor , m_reducer) , arg_result.data() );
       return ;
     }
 
     m_team_begin = UseShflReduction?0:cuda_single_inter_block_reduce_scan_shmem<false,FunctorType,WorkTag>( arg_functor , m_team_size );
     m_shmem_begin = sizeof(double) * ( m_team_size + 2 );
-    m_shmem_size = arg_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( arg_functor , m_team_size );
-    m_scratch_ptr[1] = cuda_resize_scratch_space(static_cast<std::int64_t>(m_scratch_size[1])*(static_cast<std::int64_t>(Cuda::concurrency()/(m_team_size*m_vector_size))));
+    m_shmem_size = m_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( arg_functor , m_team_size );
     m_scratch_size[0] = m_shmem_size;
-    m_scratch_size[1] = arg_policy.scratch_size(1,m_team_size);
+    m_scratch_size[1] = m_policy.scratch_size(1,m_team_size);
+    m_scratch_ptr[1] = m_team_size<=0?NULL:cuda_resize_scratch_space(static_cast<std::int64_t>(m_scratch_size[1])*(static_cast<std::int64_t>(Cuda::concurrency()/(m_team_size*m_vector_size))));
```

Comment:
@crtrott if any more changes are made could `NULL` be replaced by `nullptr`? Otherwise we can update later

Reply:
we leave it for now